### ISSUE TITLE
ag (via helm or projectile) does not work by default 

### ISSUE THREAD

**Issue by noelweichbrodt** [syl20bnr/spacemacs#7868]
#### Description :octocat:
ag does not work by default (cf comments in #3141)
#### Reproduction guide :beetle:
- Start Emacs
- M-m s a f
- use provided defaults for files and any pattern

*Observed behaviour:* :eyes: :broken_heart:
`Package 'ag' is not available`
` (file-error \"Searching for program\" \"No such file or directory\" \"ag\")")
`
*Expected behaviour:* :heart: :smile:
ag results buffer appears

#### System Info :computer:
- OS: darwin
- Emacs: 25.1.1
- Spacemacs: 0.200.5
- Spacemacs branch: master (rev. 664ba6a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(yaml helm auto-completion emacs-lisp git markdown org python clojure)
```
- System configuration features: NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES

#### Backtrace :paw_prints
```
Error running timer: (error "In ‘Search at ~/’ source: `helm-ag--do-ag-candidate-process`
(file-error \"Searching for program\" \"No such file or directory\" \"ag\")")
```

**Comment by bmag** [comment]
First, your layers list shows `ag`, but Spacemacs doesn't contain such a layer. Is it a personal layer? What does it contain?

Second, backtrace is relevant, please provide one (see [Debugging](https://github.com/syl20bnr/spacemacs/wiki/Debugging)).

Third, Spacemacs isn't supposed to depend on `ag` package, so not sure where you're getting that error from.

**Comment by noelweichbrodt** [comment]
Regarding 1), apologies, I forgot to take that layer out. You'll have to take my word it has no effect (and I've removed it from the report). I've updated the report with 2). For 3) well…I agree with you :)

**Comment by hoff2** [comment]
I'm just chiming in here to say I'm experiencing this same issue.

**Comment by hoff2** [comment]
Adding ag to the dotspacemacs-additional-packages in .spacemacs seems to have resolved this for me.

**Comment by bmag** [comment]
> Package 'ag' is not available

That points that `ag.el` is missing from your system.

> Error running timer: (error "In ‘Search at ~/’ source: `helm-ag--do-ag-candidate-process`
(file-error \\"Searching for program\\" \\"No such file or directory\\" \\"ag\\")")

That points that the error is in `helm-ag--do-ag-candidate-process`, probably in 
```elisp
(apply #'start-file-process "helm-do-ag" nil cmd-args)
```
inside of: ([source](https://github.com/syohex/emacs-helm-ag/blob/master/helm-ag.el#L990-L1009))
```elisp
(defun helm-ag--do-ag-candidate-process ()
  (let* ((non-essential nil)
         (default-directory (or helm-ag--default-directory
                                helm-ag--last-default-directory
                                default-directory))
         (cmd-args (helm-ag--construct-do-ag-command helm-pattern)))
    (when cmd-args
      (let ((proc (apply #'start-file-process "helm-do-ag" nil cmd-args)))
        (setq helm-ag--last-query helm-pattern
              helm-ag--last-command cmd-args
              helm-ag--ignore-case (helm-ag--ignore-case-p cmd-args helm-pattern)
              helm-ag--last-default-directory default-directory)
        (prog1 proc
          (set-process-sentinel
           proc
           (lambda (process event)
             (helm-process-deferred-sentinel-hook
              process event (helm-default-directory))
             (when (string= event "finished\n")
               (helm-ag--do-ag-propertize helm-input)))))))))
```

Those errors don't seem to be related. I don't know what's the matter here, unfortunately... possibly something in your environment... (does `ag` reside in a directory that isn't in a path that `start-file-process` respects? did it work in the past?)

**Comment by lvh** [comment]
@bmag Same issue here. ag is installed via homebrew; lives in `/usr/local/bin/ag`. I think your estimation is accurate; helm-ag by itself works fine.

I'm using fish as a default shell, I don't know if @hoff and @noelweichbrodt are, or if that matters.

**Comment by lvh** [comment]
To be clear: the error I see is `projectile-ag: Package ’ag’ is not available`.

**Comment by lvh** [comment]
@noelweichbrodt @bmag Are you, perchance, using hybrid mode and using <kbd>C-p s s</kbd>? It appears thats' bound to projectile-ag, which is different from the thing Spacemacs wants to do (<kbd>SPC s p</kbd> or directly via e.g. <kbd>SPC s t p</kbd>)

**Comment by noelweichbrodt** [comment]
@lvh thanks for the suggestions. Using holy-mode, not hybrid-mode, and using bash.

After `brew install ag` (into `/usr/local/bin/ag`) and restarting spacemacs, `SPC s a f` now works as expected (using helm-ag). `SPC s t p` still results in 
> Error running timer: (error "In ‘Search at ~/’ source: ‘helm-ag--do-ag-candidate-process’ 
> (file-error \"Searching for program\" \"No such file or directory\" \"pt\")")

**Comment by bmag** [comment]
@lvh I use hybrid style and `SPC /`, which is the same as `SPC s p`, which is the same as `SPC s a p` or `SPC s t p` or \<same thing for the other search tools\>. I don't use `C-p s s`. I suggest for everyone that you use `SPC /` (or `SPC s p`), `SPC s f` and `SPC s b` instead of `SPC s .. p`, `SPC s .. f` and `SPC s .. b`, because 1) easier to type and 2) will choose the search program automatically depends on what you have installed on the system.

@noelweichbrodt so for you, the problem was that you didn't have the `ag` binary installed on your system? That's possibly the same issue with `pt`. However, I'm not sure why you'd want to use `pt` if `ag` already works.

**Comment by bmag** [comment]
Look [here](https://github.com/bbatsov/projectile/blob/master/projectile.el#L2393-L2405). `projectile` offers the command `projectile-ag`, but it only works if you have `ag` package installed. Spacemacs doesn't use this command, instead we have our own command that combines `projectile` and `helm-ag` (not `ag`) packages. `C-c p s s` (what is referred to as `C-p s s` above) is `projectile`'s default binding, it isn't part of Spacemacs bindings. It seems that the two issues in this report have been resolved, so I'm closing it. Please report if there are still problems.

To summarize:
If the error is `(file-error \"Searching for program\" \"No such file or directory\" \"ag\")")`: You don't have the `ag` binary installed, or you do but Emacs can't find it. If Emacs can't find it, then it's a path issue. To solve this issue, install the `ag` binary, for example via `brew` as @noelweichbrodt did.

If the error is `Package 'ag' is not available`: You're using `C-c p s s`, which isn't officially supported. Either use `SPC /` instead (`M-m /` for Emacs style users), or install the package `ag` as an additional package as @hoff2 did.

**Comment by hoff2** [comment]
So I think the issue here is that we both came into spacemacs having committed `C-c p s s` to muscle memory from prior usage (either of `projectile-ag` or `helm-projectile-ag` which used the same key binding), and found ourselves surprised by the fact that the key binding still brings up a prompt, but then fails with a error message. Perhaps removing the default key binding would help.

Thanks for the help @bmag. I'll switch to `SPC /`.

**Comment by chenyangguang** [comment]
@hoff2 , but it fail in `dump-jump-go` to a variable or function defined.
